src/librbd/api/Mirror.cc

```diff
@@ -2146,6 +2434,1499 @@ void Mirror<I>::image_snapshot_create(I *ictx, uint32_t flags,
+                         std::vector<uint64_t> &quiesce_requests,
+                         cls::rbd::MirrorSnapshotState state,
+                         std::set<std::string> *mirror_peer_uuids,
+                         uint32_t flags) {
+  CephContext *cct = (CephContext *)group_ioctx.cct();
+  ldout(cct, 20) << dendl;
+
+  uint64_t internal_flags = 0;
+  int r = librbd::util::snap_create_flags_api_to_internal(cct, flags,
+                                                          &internal_flags);
+  if (r < 0) {
+    return r;
+  }
+
+  auto ns = group_ioctx.get_namespace();
+  group_ioctx.set_namespace("");
+  std::vector<cls::rbd::MirrorPeer> peers;
+  r = cls_client::mirror_peer_list(&group_ioctx, &peers);
+  if (r < 0) {
+    lderr(cct) << "error reading mirror peers: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+  group_ioctx.set_namespace(ns);
+
+  for (auto &peer : peers) {
+    if (peer.mirror_peer_direction == cls::rbd::MIRROR_PEER_DIRECTION_RX) {
+      continue;
+    }
+    mirror_peer_uuids->insert(peer.uuid);
+  }
+
+  if (mirror_peer_uuids->empty()) {
+    lderr(cct) << "no mirror tx peers configured for the pool" << dendl;
+    return -EINVAL;
+  }
+  r = open_group_images(group_ioctx, group_id, image_ctxs);
+  if (r < 0) {
+    return r;
+  }
+
+  group_snap->snapshot_namespace = cls::rbd::GroupSnapshotNamespaceMirror{
+                                     state, *mirror_peer_uuids, {}, {}};
+
+  for (auto image_ctx: *image_ctxs) {
+    group_snap->snaps.emplace_back(image_ctx->md_ctx.get_id(), image_ctx->id,
+                                   CEPH_NOSNAP);
+  }
+
+  int ret_code = 0;
+  std::vector<C_SaferCond*> on_finishes(image_ctxs->size(), nullptr);
+  std::string group_header_oid = librbd::util::group_header_name(group_id);
+  r = cls_client::group_snap_set(&group_ioctx, group_header_oid, *group_snap);
+  if (r < 0) {
+    lderr(cct) << "failed to set group snapshot metadata: " << cpp_strerror(r)
+               << dendl;
+    ret_code = r;
+    goto remove_record;
+  }
+
+  if ((internal_flags & SNAP_CREATE_FLAG_SKIP_NOTIFY_QUIESCE) == 0) {
+    NoOpProgressContext prog_ctx;
+    r = util::notify_quiesce(*image_ctxs, prog_ctx, &quiesce_requests);
+    if (r < 0 &&
+        (internal_flags & SNAP_CREATE_FLAG_IGNORE_NOTIFY_QUIESCE_ERROR) == 0) {
+      ret_code = r;
+      goto remove_record;
+    }
+  }
+  ldout(cct, 20) << "Requesting exclusive locks for images" << dendl;
+
+  C_SaferCond* on_finish;
+  for (auto ictx: *image_ctxs) {
+    std::shared_lock owner_lock{ictx->owner_lock};
+    if (ictx->exclusive_lock != nullptr) {
+      ictx->exclusive_lock->block_requests(-EBUSY);
+    }
+  }
+  for (size_t i = 0; i < image_ctxs->size(); ++i) {
+    ImageCtx *ictx = (*image_ctxs)[i];
+    std::shared_lock owner_lock{ictx->owner_lock};
+
+    on_finish = new C_SaferCond;
+    if (ictx->exclusive_lock != nullptr) {
+      ictx->exclusive_lock->acquire_lock(on_finish);
```

Comment:
`C_SaferCond` object is leaked here if `ictx->exclusive_lock == nullptr`.  The allocation was likely meant to be performed in the branch:
```suggestion
    if (ictx->exclusive_lock != nullptr) {
      on_finish = new C_SaferCond;
      ictx->exclusive_lock->acquire_lock(on_finish);
```

Reply:
Addressed, thanks Ilya